runtimes/common/src/bonded_currencies/mod.rs

```diff
@@ -0,0 +1,305 @@
+		NativeAndForeignAssets::<Left, Right, Criterion, AssetKind, AccountId>::mint_into(asset, who, amount)
+	}
+	fn burn_from(
+		asset: Self::AssetId,
+		who: &AccountId,
+		amount: Self::Balance,
+		precision: Precision,
+		force: Fortitude,
+	) -> Result<Self::Balance, DispatchError> {
+		NativeAndForeignAssets::<Left, Right, Criterion, AssetKind, AccountId>::burn_from(
+			asset, who, amount, precision, force,
+		)
+	}
+	fn shelve(asset: Self::AssetId, who: &AccountId, amount: Self::Balance) -> Result<Self::Balance, DispatchError> {
+		NativeAndForeignAssets::<Left, Right, Criterion, AssetKind, AccountId>::shelve(asset, who, amount)
+	}
+	fn restore(asset: Self::AssetId, who: &AccountId, amount: Self::Balance) -> Result<Self::Balance, DispatchError> {
+		NativeAndForeignAssets::<Left, Right, Criterion, AssetKind, AccountId>::restore(asset, who, amount)
+	}
+	fn transfer(
+		asset: Self::AssetId,
+		source: &AccountId,
+		dest: &AccountId,
+		amount: Self::Balance,
+		preservation: Preservation,
+	) -> Result<Self::Balance, DispatchError> {
+		NativeAndForeignAssets::<Left, Right, Criterion, AssetKind, AccountId>::transfer(
+			asset,
+			source,
+			dest,
+			amount,
+			preservation,
+		)
+	}
+
+	fn set_balance(asset: Self::AssetId, who: &AccountId, amount: Self::Balance) -> Self::Balance {
+		NativeAndForeignAssets::<Left, Right, Criterion, AssetKind, AccountId>::set_balance(asset, who, amount)
+	}
+}
+
+impl<
+		Left: fungible::Inspect<AccountId>
+			+ AccountTouch<(), AccountId, Balance = <Left as fungible::Inspect<AccountId>>::Balance>,
+		Right: fungibles::Inspect<AccountId>
+			+ AccountTouch<Right::AssetId, AccountId, Balance = <Left as fungible::Inspect<AccountId>>::Balance>,
+		Criterion: Convert<AssetKind, Either<(), Right::AssetId>>,
+		AssetKind: AssetIdTraits,
+		AccountId,
+	> AccountTouch<AssetKind, AccountId> for WrapperNativeAndForeignAssets<Left, Right, Criterion, AssetKind, AccountId>
+{
+	type Balance = <Left as fungible::Inspect<AccountId>>::Balance;
+
+	fn deposit_required(asset: AssetKind) -> Self::Balance {
+		NativeAndForeignAssets::<Left, Right, Criterion, AssetKind, AccountId>::deposit_required(asset)
+	}
+
+	fn should_touch(asset: AssetKind, who: &AccountId) -> bool {
+		NativeAndForeignAssets::<Left, Right, Criterion, AssetKind, AccountId>::should_touch(asset, who)
+	}
+
+	fn touch(asset: AssetKind, who: &AccountId, depositor: &AccountId) -> DispatchResult {
+		NativeAndForeignAssets::<Left, Right, Criterion, AssetKind, AccountId>::touch(asset, who, depositor)
+	}
+}
+
+impl<
+		Left: fungible::Inspect<AccountId>,
+		Right: Inspect<AccountId, Balance = <Left as fungible::Inspect<AccountId>>::Balance>,
+		Criterion: Convert<AssetKind, Either<(), Right::AssetId>>,
+		AssetKind: AssetIdTraits,
+		AccountId,
+	> Inspect<AccountId> for WrapperNativeAndForeignAssets<Left, Right, Criterion, AssetKind, AccountId>
+{
+	fn decimals(asset: Self::AssetId) -> u8 {
+		match Criterion::convert(asset) {
+			Left(()) => 15u8,
```

Comment:
Should I move this into const?

Reply:
It is not used multiple time

Reply:
Also, shouldn't this use the decimal amount specified for the balances pallet? Or why are we returning `15` by default?

Reply:
if the balances pallet allows us to read the native currency's denomination, it would indeed make sense to do so

Reply:
The balances pallet is not exposing such information. The denomination are metadata, set in the chain spec. At runtime level there are unknown. 
Same for symbol and name. 

I have moved the denomination into a constant in: [bc4dbae](https://github.com/KILTprotocol/kilt-node/pull/818/commits/bc4dbae37ee31eca96bb419b7174c20e89b00ba1). 

I think symbol and name are not necessary. They are not used in the pallet, and we will not need them in the runtime api.

Reply:
NVM we need them in the runtime api: [8330fa8](https://github.com/KILTprotocol/kilt-node/pull/818/commits/8330fa81335a63e6cf78d77ccae89ab28049ade1)

Reply:
I would suggest you move the value of decimals for peregrine into the peregrine runtime crate, and use that also when generating the chainspec in the binary. Then, you can use the same const here.

Reply:
Sorry but your commit history seems to be a mess and all the commits you link to give a 404 error.

Reply:
I changed a bit the implementation. i introduced another generic in the type to fetch the metadata: [0dcf5b3](https://github.com/KILTprotocol/kilt-node/pull/818/commits/0dcf5b38aaf642cb694d3e8cc7cb222218a091ea) 

It's adding a bit more complexity. Let me know what you think

Reply:
yeah sorry I was thinking rebasing would be a good idea...